File: rb/test/api_entities_package_pipeline_entity_test.rb
# ApiEntitiesPackagePipeline entity test

require "minitest/autorun"
require "json"
require_relative "../Gitlab_sdk"
require_relative "runner"

class ApiEntitiesPackagePipelineEntityTest < Minitest::Test
  def test_create_instance
    testsdk = GitlabSDK.test(nil, nil)
    ent = testsdk.ApiEntitiesPackagePipeline(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = api_entities_package_pipeline_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["load"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "api_entities_package_pipeline." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set GITLAB_TEST_API_ENTITIES_PACKAGE_PIPELINE_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    api_entities_package_pipeline_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.api_entities_package_pipeline")))
    api_entities_package_pipeline_ref01_data = nil
    if api_entities_package_pipeline_ref01_data_raw.length > 0
      api_entities_package_pipeline_ref01_data = Helpers.to_map(api_entities_package_pipeline_ref01_data_raw[0][1])
    end

    # LOAD
    api_entities_package_pipeline_ref01_ent = client.ApiEntitiesPackagePipeline(nil)
    api_entities_package_pipeline_ref01_match_dt0 = {
      "id" => api_entities_package_pipeline_ref01_data["id"],
    }
    api_entities_package_pipeline_ref01_data_dt0_loaded = api_entities_package_pipeline_ref01_ent.load(api_entities_package_pipeline_ref01_match_dt0, nil)
    api_entities_package_pipeline_ref01_data_dt0_load_result = Helpers.to_map(api_entities_package_pipeline_ref01_data_dt0_loaded)
    assert !api_entities_package_pipeline_ref01_data_dt0_load_result.nil?
    assert_equal api_entities_package_pipeline_ref01_data_dt0_load_result["id"], api_entities_package_pipeline_ref01_data["id"]

  end
end

def api_entities_package_pipeline_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "api_entities_package_pipeline", "ApiEntitiesPackagePipelineTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = GitlabSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["api_entities_package_pipeline01", "api_entities_package_pipeline02", "api_entities_package_pipeline03", "project01", "project02", "project03", "package01", "package02", "package03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["GITLAB_TEST_API_ENTITIES_PACKAGE_PIPELINE_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "GITLAB_TEST_API_ENTITIES_PACKAGE_PIPELINE_ENTID" => idmap,
    "GITLAB_TEST_LIVE" => "FALSE",
    "GITLAB_TEST_EXPLAIN" => "FALSE",
    "GITLAB_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["GITLAB_TEST_API_ENTITIES_PACKAGE_PIPELINE_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["GITLAB_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["GITLAB_APIKEY"],
      },
      extra || {},
    ])
    client = GitlabSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["GITLAB_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["GITLAB_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end

File: go/test/api_entities_package_pipeline_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/gitlab-sdk/go"
	"github.com/voxgig-sdk/gitlab-sdk/go/core"

	vs "github.com/voxgig-sdk/gitlab-sdk/go/utility/struct"
)

func TestApiEntitiesPackagePipelineEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.ApiEntitiesPackagePipeline(nil)
		if ent == nil {
			t.Fatal("expected non-nil ApiEntitiesPackagePipelineEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := api_entities_package_pipelineBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "api_entities_package_pipeline." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set GITLAB_TEST_API_ENTITIES_PACKAGE_PIPELINE_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		apiEntitiesPackagePipelineRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.api_entities_package_pipeline", setup.data)))
		var apiEntitiesPackagePipelineRef01Data map[string]any
		if len(apiEntitiesPackagePipelineRef01DataRaw) > 0 {
			apiEntitiesPackagePipelineRef01Data = core.ToMapAny(apiEntitiesPackagePipelineRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = apiEntitiesPackagePipelineRef01Data

		// LOAD
		apiEntitiesPackagePipelineRef01Ent := client.ApiEntitiesPackagePipeline(nil)
		apiEntitiesPackagePipelineRef01MatchDt0 := map[string]any{
			"id": apiEntitiesPackagePipelineRef01Data["id"],
		}
		apiEntitiesPackagePipelineRef01DataDt0Loaded, err := apiEntitiesPackagePipelineRef01Ent.Load(apiEntitiesPackagePipelineRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		apiEntitiesPackagePipelineRef01DataDt0LoadResult := core.ToMapAny(apiEntitiesPackagePipelineRef01DataDt0Loaded)
		if apiEntitiesPackagePipelineRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if apiEntitiesPackagePipelineRef01DataDt0LoadResult["id"] != apiEntitiesPackagePipelineRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func api_entities_package_pipelineBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "api_entities_package_pipeline", "ApiEntitiesPackagePipelineTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read api_entities_package_pipeline test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse api_entities_package_pipeline test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"api_entities_package_pipeline01", "api_entities_package_pipeline02", "api_entities_package_pipeline03", "project01", "project02", "project03", "package01", "package02", "package03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("GITLAB_TEST_API_ENTITIES_PACKAGE_PIPELINE_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"GITLAB_TEST_API_ENTITIES_PACKAGE_PIPELINE_ENTID": idmap,
		"GITLAB_TEST_LIVE":      "FALSE",
		"GITLAB_TEST_EXPLAIN":   "FALSE",
		"GITLAB_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["GITLAB_TEST_API_ENTITIES_PACKAGE_PIPELINE_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["GITLAB_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["GITLAB_APIKEY"],
			},
			extra,
		})
		client = sdk.NewGitlabSDK(core.ToMapAny(mergedOpts))
	}

	live := env["GITLAB_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["GITLAB_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
